torch/csrc/jit/tensorexpr/cuda_codegen.cpp

```diff
@@ -301,43 +264,98 @@ static bool CheckEqual(const Expr* lhs, const Expr* rhs) {
   return immediateEquals(diff_s.node(), 0);
 }
 
-// Identify the pattern: a[e1] = a[e1] + e2.
-static bool isAtomicAdd(const Store* v, const Expr** atomic_add_value) {
-  ScalarType dtype = v->value()->dtype().scalar_type();
-  if (dtype != ScalarType::Float && dtype != ScalarType::Double) {
-    return false;
-  }
-  const Add* add_v = dynamic_cast<const Add*>(v->value());
-  if (!add_v) {
-    return false;
-  }
-  const Load* load_v = dynamic_cast<const Load*>(add_v->lhs());
-  if (!load_v) {
-    return false;
-  }
-  if (v->base_handle() != load_v->base_handle()) {
-    return false;
-  }
-  if (v->indices().empty() && load_v->indices().empty()) {
-    return false;
-  }
-  bool index_equal = CheckEqual(v->flat_index(), load_v->flat_index());
-  if (index_equal) {
-    *atomic_add_value = add_v->rhs();
+class AtomicAddFuser : public IRMutator {
+ public:
+  AtomicAddFuser(
+      const std::unordered_set<const Var*>& thread_local_bufs,
+      const GPUMetaVarRewriter* metavars)
+      : thread_local_bufs_(thread_local_bufs) {
+    size_t DIMS = 3;
+
+    const std::vector<const Expr*>& block_extents =
+        metavars->gpu_block_extents();
+    const std::vector<const Var*>& block_vars = metavars->gpu_block_vars();
+    for (size_t i = 0; i < block_extents.size(); ++i) {
+      MetaVarExtent extent{block_extents[i], false};
+      if (extent.expr->isConstant() && immediateEquals(extent.expr, 1)) {
+        extent.trivial = true;
+      } else {
+        nontrivial_metavars_.insert(block_vars[i]);
+      }
+      metavars_[block_vars[i]] = extent;
+    }
+
+    const std::vector<const Expr*>& thread_extents =
+        metavars->gpu_thread_extents();
+    const std::vector<const Var*>& thread_vars = metavars->gpu_thread_vars();
+    for (size_t i = 0; i < thread_extents.size(); ++i) {
+      MetaVarExtent extent{thread_extents[i], false};
+      if (extent.expr->isConstant() && immediateEquals(extent.expr, 1)) {
```

Comment:
Minor: this seems to be idiomatic in our code base now. Maybe we should have an overloaded "immediateEquals" that does this now?

Reply:
Sure, makes sense - in a follow up though.